sklearn/metrics/_dist_metrics.pyx.tp

```diff
@@ -2785,6 +2789,60 @@ cdef class HaversineDistance{{name_suffix}}(DistanceMetric{{name_suffix}}):
 
         return (sin_0 * sin_0 + cos(x1_0) * cos(x2_0) * sin_1 * sin_1)
 
+#------------------------------------------------------------
+# Arccos Distance
+#  D(x, y) = arccos(dot(x, y) / (|x| * |y|)) / PI
+# This is the metric version of cosine distance, i.e. the metric
+# most suitable when you wish to use cosine distance.
+
+cdef class ArccosDistance{{name_suffix}}(DistanceMetric{{name_suffix}}):
+    cdef inline {{INPUT_DTYPE_t}} cosine_similarity(
+        self,
+        const {{INPUT_DTYPE_t}}* x1,
+        const {{INPUT_DTYPE_t}}* x2,
+        ITYPE_t size
+    ) nogil except -1:
+        cdef DTYPE_t d = _dot(size, x1, 1, x2, 1)
+        cdef DTYPE_t norm1 = _dot(size, x1, 1, x1, 1)
+        cdef DTYPE_t norm2 = _dot(size, x2, 1, x2, 1)
+        return d / sqrt(norm1 * norm2)
```

Comment:
To answer your comment regarding benchmarking various versions of this method (i.e. https://github.com/scikit-learn/scikit-learn/pull/24106#discussion_r944395136), I think the best is to perform a standalone comparison without scikit-learn:

To do this, I recommend that you use:
 - [Cython in a Notebook using annotation mode](https://cython.readthedocs.io/en/latest/src/quickstart/build.html#using-the-jupyter-notebook) to understand how to minimize the overhead of CPython (there must be some in `def` functions but none in `cdef` function)
 - several `cdef` functions for variations of `cosine_similarity` which define the core of the computations
 - a Python `def` function as a front-end for each `cdef` function. Such function must just transfer the `x1` and `x2` numpy arrays' buffers' pointers to those `cdef` functions.
 - the `%timeit` magic on those Python `def` functions to assert performance of `cdef` functions. You better use large `x1` and `x2` arrays not to have the interaction with CPython have the runtime overhead majority.
 - report the notebook somewhere publicly (e.g. a public repository on GitHub) and link this notebook in a comment here with conclusions. This way reviewers can rerun it and better suggestion on top of your work.